domain/src/main/java/org/oppia/android/domain/oppialogger/analytics/CpuPerformanceSnapshotter.kt

```diff
@@ -0,0 +1,206 @@
+package org.oppia.android.domain.oppialogger.analytics
+
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.channels.SendChannel
+import kotlinx.coroutines.channels.actor
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import org.oppia.android.app.model.ScreenName
+import org.oppia.android.util.logging.ConsoleLogger
+import org.oppia.android.util.logging.ExceptionLogger
+import org.oppia.android.util.logging.performancemetrics.PerformanceMetricsAssessor
+import org.oppia.android.util.logging.performancemetrics.PerformanceMetricsAssessor.AppIconification
+import org.oppia.android.util.logging.performancemetrics.PerformanceMetricsAssessor.CpuSnapshot
+
+/**
+ * Snapshotter that gracefully and sequentially logs CPU usage across foreground and background
+ * moves of the application.
+ */
+class CpuPerformanceSnapshotter(
+  private val backgroundCoroutineDispatcher: CoroutineDispatcher,
+  private val performanceMetricsLogger: PerformanceMetricsLogger,
+  private val initialIconification: AppIconification,
+  private val consoleLogger: ConsoleLogger,
+  private val exceptionLogger: ExceptionLogger,
+  private val foregroundCpuLoggingTimePeriodMillis: Long,
+  private val backgroundCpuLoggingTimePeriodMillis: Long,
+  private val performanceMetricsAssessor: PerformanceMetricsAssessor
+) {
+
+  private val commandQueue by lazy { createCommandQueueActor() }
+
+  /** Updates the current [AppIconification] in accordance with the app's state changes. */
+  fun updateAppIconification(newIconification: AppIconification) {
+    sendSwitchIconificationCommand(newIconification)
+  }
+
+  private fun createCommandQueueActor(): SendChannel<CommandMessage> {
+    var currentIconification = initialIconification
+    var previousSnapshot =
+      performanceMetricsAssessor.computeCpuSnapshotAtCurrentTime()
+    var switchIconificationCount = 0
+    val coroutineScope = CoroutineScope(backgroundCoroutineDispatcher)
+    return coroutineScope.actor<CommandMessage>(capacity = Channel.UNLIMITED) {
+      for (message in channel) {
+        when (message) {
+          is CommandMessage.SwitchIconification -> {
+            ++switchIconificationCount
+            // Since there's a switch in the current iconification of the app, we'd cut short the
+            // existing delay and log the current CPU usage relative to the previously logged one
+            // without this explicit log command.
+            val relativeCpuUsage = performanceMetricsAssessor.getRelativeCpuUsage(
+              previousSnapshot,
+              performanceMetricsAssessor.computeCpuSnapshotAtCurrentTime()
+            )
+            relativeCpuUsage?.let { value ->
+              sendLogSnapshotDiffCommand(
+                value,
+                currentIconification
+              )
+            }
```

Comment:
```suggestion
            performanceMetricsAssessor.getRelativeCpuUsage(
              previousSnapshot,
              performanceMetricsAssessor.computeCpuSnapshotAtCurrentTime()
            )?.let { relativeCpuUsage ->
              sendLogSnapshotDiffCommand(relativeCpuUsage, currentIconification)
            }
```

Can be compacted a bit.

Reply:
thanks for the suggestion! -- done.